src/track/track_gps_l2cm.c

```diff
@@ -0,0 +1,614 @@
+
+  SETTING(L2CM_TRACK_SETTING_SECTION, "cn0_drop",
+          track_cn0_drop_thres, TYPE_FLOAT);
+
+  SETTING(L2CM_TRACK_SETTING_SECTION, "alias_detect",
+          use_alias_detection, TYPE_BOOL);
+
+  for (u32 i = 0; i < NUM_GPS_L2CM_TRACKERS; i++) {
+    gps_l2cm_trackers[i].active = false;
+    gps_l2cm_trackers[i].data = &gps_l2cm_tracker_data[i];
+  }
+
+  tracker_interface_register(&tracker_interface_list_element_gps_l2cm);
+}
+
+/** Do L1C/A to L2 CM handover.
+ *
+ * The condition for the handover is the availability of bitsync on L1 C/A
+ *
+ * \param sample_count NAP sample count
+ * \param sat L1C/A Satellite ID
+ * \param nap_channel Associated NAP channel
+ * \param code_phase L1CA code phase [chips]
+ */
+void do_l1ca_to_l2cm_handover(u32 sample_count,
+                              u16 sat,
+                              u8 nap_channel,
+                              float code_phase)
+{
+  /* First, get L2C capability for the SV from NDB */
+  u32 l2c_cpbl;
+  // TODO: uncomment this as soon as NDB gets available
+  // ndb_gps_l2cm_l2c_cap_read(&l2c_cpbl);
+
+  // TODO: remove this as soon as NDB gets available
+  // GPS PRNs with L2C capability:
+  // 01 03 05 06 07 08 09 10 12 15 17 24 25 26 27 29 30 31 32
+  // as per http://tinyurl.com/zj5q62h
+  l2c_cpbl = (u32)1 << (1 - 1);
+  l2c_cpbl |= (u32)1 << (3 - 1);
+  l2c_cpbl |= (u32)1 << (5 - 1);
+  l2c_cpbl |= (u32)1 << (6 - 1);
+  l2c_cpbl |= (u32)1 << (7 - 1);
+  l2c_cpbl |= (u32)1 << (8 - 1);
+  l2c_cpbl |= (u32)1 << (9 - 1);
+  l2c_cpbl |= (u32)1 << (10 - 1);
+  l2c_cpbl |= (u32)1 << (12 - 1);
+  l2c_cpbl |= (u32)1 << (15 - 1);
+  l2c_cpbl |= (u32)1 << (17 - 1);
+  l2c_cpbl |= (u32)1 << (24 - 1);
+  l2c_cpbl |= (u32)1 << (25 - 1);
+  l2c_cpbl |= (u32)1 << (26 - 1);
+  l2c_cpbl |= (u32)1 << (27 - 1);
+  l2c_cpbl |= (u32)1 << (29 - 1);
+  l2c_cpbl |= (u32)1 << (30 - 1);
+  l2c_cpbl |= (u32)1 << (31 - 1);
+  l2c_cpbl |= (u32)1 << (32 - 1);
+
+  /* compose SID: same SV, but code is L2 CM */
+  gnss_signal_t sid = construct_sid(CODE_GPS_L2CM, sat);
+
+  if (!tracking_startup_ready(sid)) {
+    return; /* L2C signal from the SV is already in track */
+  }
+
+  if (0 == (l2c_cpbl & ((u32)1 << (sat - 1)))) {
+    log_info_sid(sid, "SV does not support L2C signal");
+    return;
+  }
+
+  if ((code_phase < 0) ||
+      ((code_phase > 0.5) && (code_phase < (GPS_L1CA_CHIPS_NUM - 0.5)))) {
+    log_warn_sid(sid, "Unexpected L1C/A to L2C handover code phase: %f",
+                 code_phase);
+    return;
+  }
+
+  if (code_phase > (GPS_L1CA_CHIPS_NUM - 0.5)) {
+    code_phase = 2 * GPS_L2CM_CHIPS_NUM - (GPS_L1CA_CHIPS_NUM - code_phase);
+  }
+
+  /* recalculate doppler freq for L2 from L1*/
+  double carrier_freq = tracking_channel_carrier_freq_get(nap_channel) *
+                        GPS_L2_HZ / GPS_L1_HZ;
+
+  /* get initial cn0 from parent L1 channel */
+  float cn0_init = tracking_channel_cn0_get(nap_channel);
+
+  /* The best elevation estimation could be retrieved by calling
```

Comment:
FYI I think elevation should go in the NDB or somewhere other than the tracking channels eventually. Right now it is computed by the solution thread periodically and written back to the tracking channels, which which is clumsy. I'm mostly just trying to maintain the architecture where `track_gps_l1ca.c` and similar are slaves to `track.c` and only access data via the API in `track_api.h`